internal/app/upstream/client.go

```diff
@@ -55,11 +73,184 @@ type Response struct {
 //
 // The method does not block until the underlying connection is up.
 // Returns immediately and connecting the server happens in background
-// TODO: pass retry/timeout configurations
-func NewClient(ctx context.Context, url string) (Client, error) {
-	return &client{}, nil
+func NewClient(ctx context.Context, url string, callOptions CallOptions) (Client, error) {
+	// TODO: configure grpc options.https://github.com/envoyproxy/xds-relay/issues/55
+	conn, err := grpc.Dial(url, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	ldsClient := v2.NewListenerDiscoveryServiceClient(conn)
+	rdsClient := v2.NewRouteDiscoveryServiceClient(conn)
+	edsClient := v2.NewEndpointDiscoveryServiceClient(conn)
+	cdsClient := v2.NewClusterDiscoveryServiceClient(conn)
+
+	go shutDown(ctx, conn)
+
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}, nil
+}
+
+func (m *client) OpenStream(request *v2.DiscoveryRequest) (<-chan *v2.DiscoveryResponse, chan bool, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	var stream grpc.ClientStream
+	var err error
+	switch request.GetTypeUrl() {
+	case ListenerTypeURL:
+		stream, err = m.ldsClient.StreamListeners(ctx)
+	case ClusterTypeURL:
+		stream, err = m.cdsClient.StreamClusters(ctx)
+	case RouteTypeURL:
+		stream, err = m.rdsClient.StreamRoutes(ctx)
+	case EndpointTypeURL:
+		stream, err = m.edsClient.StreamEndpoints(ctx)
+	default:
+		cancel()
+		return nil, nil, &UnsupportedResourceError{TypeURL: request.GetTypeUrl()}
+	}
+
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+
+	signal := make(chan *version, 1)
+	// The xds protocol https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack
+	// specifies that the first request be empty nonce and empty version.
+	// The origin server will respond with the latest version.
+	signal <- &version{nonce: "", version: ""}
+
+	done := make(chan bool, 1)
+	response := make(chan *v2.DiscoveryResponse, 1)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
```

Comment:
What's the reason that we wait on both recv and send to fail (in the case of an error) or succeed (in the case of a downstream signal to terminate), rather than just either/or? Can this just be `wg.Add(1)`?

Reply:
The waitgroup guards the channels which are used in both send and recv goroutines. We need both goroutines to exit before we close the channels. We could block on both channels separately, but both in error and success cases we need both channels to be closed and hence they are together in one group.